spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionedRDD.scala

```diff
@@ -22,15 +32,41 @@ class CassandraPartitionedRDD[T](prev: RDD[T])(implicit ct: ClassTag[T]) extends
   override def getPartitions: Array[Partition] = {
     partitioner match {
       case Some(rp: ReplicaPartitioner) => prev.partitions.map(partition => rp.getEndpointPartition(partition))
-      case _ => throw new IllegalArgumentException("CassandraPartitionedRDD hasn't been partitioned by ReplicaPartitioner. This should be impossible")
+      case _ => throw new IllegalArgumentException("CassandraPartitionedRDD hasn't been " +
+        "partitioned by ReplicaPartitioner. This should be impossible")
     }
   }
 
+  /**
+   * This method currently uses thrift to determine the local endpoint and rpc endpoints
+   * from whatever endpoint the partition belongs to. This gives us a better chance of matching
+   * the bound interface of the Spark Executor. In addition we will add the HostName and
+   * HostAddress that we get for each of these endpoints to cover all of the logical choices.
+   */
   override def getPreferredLocations(split: Partition): Seq[String] = {
+    val connector = CassandraConnector(prev.sparkContext.getConf)
+
+    val rpcToLocalAddress = connector.withCassandraClientDo { client =>
+      val ring = client.describe_local_ring(keyspace)
+      for {
+        tr <- ring
+        rpcIps = tr.rpc_endpoints.map(InetAddress.getByName)
+        localIps = tr.endpoints.map(InetAddress.getByName)
+        (rpc, local) <- (rpcIps).zip(localIps)
+      } yield (rpc, local)
```

Comment:
@RussellSpitzer can you just do 

yield (local,rpc) here and avoid the traversal below? They have the same data  just a different tuple sequence